Automattic/react-virtualized
# Add JSX output option

### For context

`react-virtualized` performs its own transpilation, adding a runtime dependency on `@babel/runtime` to implement features that are more complex than simple syntax transforms. This unfortunately means that the entire set of helpers needs to be shipped with the build, regardless of whether the features have native support.

In order to get around this, transpiling and polyfilling should be left to the consumer, and not the library, as the consumer holds all the information on target browsers, and is thus able to choose which polyfills should be shipped to particular sets of users as part of a holistic build process. For this to happen, the library would need to ship sources for the consumer to transpile as needed.

The issue is that the format these sources should be in is unclear, since JS is in a context state of flux, with projects adopting features at various stages of development. `react-virtualized` is a good example of that, adopting non-standard extensions to JS syntax like `flow`, and candidate features that were still in development, such as class properties. The goal would be to transpile to a format that all build processes would be able to understand with no changes to their Babel configs, and without requiring runtime helpers with that transpiled code itself, as that would lead us back to the same issue.

@jsnajdr proposed transpiling to a well-defined, mostly well-supported standard, such as ES2019, and to upstream these changes to the `react-virtualized` project, for all users to potentially benefit from. However, it appears that Babel no longer has those output options, having switched to a `browserslist`-based approach instead entirely. As such, we don't have a good generic solution to propose to the upstream project, and we are probably better off crafting something specific to our needs. I have forked the main repo and am preparing this PR as a result.

### Maintenance overhead

Given that `react-virtualized` appears to be in maintenance mode and that our own use of it is expected to remain stable, I don't expect that maintaining a fork will add much overhead, particularly when most of the changes are at the build (not source) level.

### This PR

Add JSX output option, and use it as the only code to be shipped.

The build is modified to perform minimal transpilation, with JSX preserved. Some transformations include converting class properties to constructor assignments, removing flow, and removing prop types.

In addition, a dependency on `dom-helpers` is removed, because it transitively added the `@babel/runtime` dependency as well. The small amount of code in use is copied to a new file in `utils/`.

A few other changes:
- Update `package.json` to point to this repo
- Add `sideEffects: false` to `package.json`, to help with tree-shaking in webpack-based consumers
- Update some documentation files to reflect the fact that this is a fork

### Next steps

Once this is merged, the goal is to publish the new package (`@automattic/react-virtualized`) publicly and switch to it in Calypso. There aren't many source files to edit, so a manual replace is probably preferable to webpack-based module rewriting.

### Testing

- Pull this branch
- Install yarn (sorry...)
- Install dependencies with `yarn`
- Build the project with `yarn build`
- Analyse the files in `dist/`

The output should be standard ES + JSX, which we can easily transpile in Calypso.
